File: src/main/scala/org/sunbird/cloud/storage/IStorageService.scala
package org.sunbird.cloud.storage

import org.sunbird.cloud.storage.Model.Blob

import scala.concurrent.{ExecutionContext, Future}

trait IStorageService {

    def uploadFolder(container: String, file: String, objectKey: String, isPublic: Option[Boolean] = Option(false), ttl: Option[Int] = None, retryCount: Option[Int] = None, attempt: Int = 1)(implicit execution: ExecutionContext) : Future[List[String]];

    /**
      *
      * Upload a file/folder to cloud
      *
      * @param container String - The container/bucket to upload the file to.
      * @param file String - The file path.
      * @param objectKey String - The destination key/path to upload the file to. If the path exists it will be overwritten.
      * @param isDirectory Option[Boolean] - Whether the file is a directory and need to upload folder recursively?
      * @param attempt
      * @param retryCount Option[Int] - Number of times the upload will be retried before failing. Defaults to global configuration "max.retries"
      * @param ttl Option[Int] - The ttl/expiry for the file. Optional and default is never expires
      * @return String - The url of the file/folder uploaded
      */
    def upload(container: String, file: String, objectKey: String, isDirectory: Option[Boolean] = Option(false),  attempt: Option[Int] = Option(1), retryCount: Option[Int] = None, ttl: Option[Int] = None): String

        
    /**
     * Put a blob in the cloud with the given content data. The difference between this and <code>upload()</code> method is that this method takes in byte array and sets is payload for the object.
     * 
     * @param container String - The container/bucket to upload the file to.
     * @param content String - The byte array of an object.
     * @param objectKey String - The destination key/path to upload the file to. If the path exists it will be overwritten.
     * @param isPublic Option[Boolean] - Whether the file should have public read access? Optional and defaults to false.
     * @param isDirectory Option[Boolean] - Whether the file is a directory and need to upload folder recursively? Optional and defaults to false.
     * @param ttl Option[Int] - The ttl/expiry for the file. Optional and default is never expires
     * @param retryCount Option[Int] - Number of times the upload will be retried before failing. Defaults to "max.retries" defined in global configuration 
     * 
     * @return String - The url of the file/folder uploaded
     */            
    def put(container: String, content: Array[Byte], objectKey: String, isPublic: Option[Boolean] = Option(false), isDirectory: Option[Boolean] = Option(false), 
            ttl: Option[Int] = None, retryCount: Option[Int] = None): String

    /**
     * Get pre-signed URL to access an object in the cloud store.
     * 
     * @param container String - The container/bucket of the file
     * @param objectKey String - The key/path of the file to pre-sign
     * @param ttl Option[Int] - The ttl/expiry for the pre-signed URL. Defaults to "max.signedurl.ttl" defined in global configuration*
     * @param permission String - The permission of pre-signed url the values are w (write), r (read). Defaults to "read".
     *
     * @return String - The pre-signed url
     */
    def getSignedURL(container: String, objectKey: String, ttl: Option[Int] = None, permission: Option[String] = Option("r")): String

    /**
     * Download file/folder from cloud storage
     * 
     * @param container String - The container/bucket of the file
     * @param objectKey String - The key/path of the file to download from
     * @param localPath String - The local destination path to download to
     * @param isDirectory Option[Boolean] - Whether the file is a directory and need to be downloaded recursively? Optional and defaults to false.
     */
    def download(container: String, objectKey: String, localPath: String, isDirectory: Option[Boolean] = Option(false))

    /**
     * Delete an object from the cloud store
     * 
     * @param container String - The container/bucket of the object.
     * @param objectKey String - The object key to delete.
     * @param isDirectory Option[Boolean] - Whether the object is a directory and need to be deleted recursively? Optional and defaults to false.
     */
    def deleteObject(container: String, objectKey: String, isDirectory: Option[Boolean] = Option(false))

    /**
     * Delete objects from the cloud store 
     * 
     * @param container String - The container/bucket of the objects
     * @param objectKeys List[(String, Boolean)] - The objects to delete. The tuple contains the object key to be deleted and whether the object is a folder so that recursive delete is triggered.
     */
    def deleteObjects(container: String, objectKeys: List[(String, Boolean)])

    /**
     * Copy objects from one container to another container or between different folders within the same container. This assumes that the credentials provided have access to both containers.
     * 
     * @param fromContainer String - The container to copy the object from.
     * @param fromKey String - The object prefix to copy from.
     * @param toContainer String - The container to copy the object to.
     * @param toKey String - The object prefix to copy to.
     * @param isDirectory Option[Boolean] - Whether the copy is a file or folder? Defaults to false i.e copy one file.
     */
    def copyObjects(fromContainer: String, fromKey: String, toContainer: String, toKey: String, isDirectory: Option[Boolean] = Option(false))


    /**
     * Remote extract a archived file on cloud storage to a given folder within the same container. 
     * 
     * @param container String - The container/bucket of the archive file.
     * @param objectKey String - The blob object archive file
     * @param toKey String - The destination folder on the bucket/container to extract to.
     */
    def extractArchive(container: String, objectKey: String, toKey: String)

    /**
     * Get the blob object details
     * 
     * @param container String - The container/bucket of the file
     * @param objectKey String - The key/path of the blob object
     * @param withPayload Option[Boolean] - Get payload as well while fetching details. Defaults to false.
     * 
     * @return Blob - The blob object.
     */
    def getObject(container: String, objectKey: String, withPayload: Option[Boolean] = Option(false)): Blob

    /**
     * List objects from cloud storage for a given prefix.
     * 
     * @param container String - The container/bucket
     * @param prefix String - The object prefix to list objects. The prefix can be folder or pattern.
     * @param withPayload Option[Boolean] - Does the listing of objects include payload as well? Defaults to false
     * 
     * @return List[Blob] - The blob objects for the given prefix.
     */
    def listObjects(container: String, prefix: String, withPayload: Option[Boolean] = Option(false)): List[Blob]

    /**
     * List object keys from cloud storage for a given prefix. Similar to <code>listObjects()</code>
     * 
     * @param container String - The container/bucket
     * @param _prefix String - The object prefix to list objects. The prefix can be folder or pattern.
     * 
     * @return List[Blob] - The blob objects for the given prefix.
     */
    def listObjectKeys(container: String, _prefix: String): List[String]

    /**
     * Search for objects for a given prefix and return only keys. Specifically used for telemetry files as the files are prefixed by sync date. 
     * By design the payload is not included in the search. If payload is required use <code>getObject()</code> method after the search is complete.
     * 
     * // TODO: Provide examples of all combinations of fromDate, toDate and delta
     * 
     * @param container String - The container/bucket.
     * @param prefix String - The prefix to search on.
     * @param fromDate Option[String] - The date to search from. Optional
     * @param toDate Option[String] - The date to search to. Optional
     * @param delta Option[Int] - The delta to search from given a from date or to date. Optional. If delta is provided and both fromDate and toDate are empty, the toDate will be defaulted to current date
     * @param pattern String - The date pattern of from and to date. Defaulst to "yyyy-MM-dd"
     * 
     * @return List[Blob] - The object keys
     */
    def searchObjects(container: String, prefix: String, fromDate: Option[String] = None, toDate: Option[String] = None, delta: Option[Int] = None, 
            pattern: String = "yyyy-MM-dd"): List[Blob]

    /**
     * Similar to <code>searchObjects()</code>. The only difference is that this method only returns the object keys instead of full blob objects.
     * 
     * @param container String - The container/bucket.
     * @param prefix String - The prefix to search on.
     * @param fromDate Option[String] - The date to search from. Optional
     * @param toDate Option[String] - The date to search to. Optional
     * @param delta Option[Int] - The delta to search from given a from date or to date. Optional. If delta is provided and both fromDate and toDate are empty, the toDate will be defaulted to current date
     * @param pattern String - The date pattern of from and to date. Defaulst to "yyyy-MM-dd"
     * 
     * @return List[String] - The object keys
     */
    def searchObjectkeys(container: String, prefix: String, fromDate: Option[String] = None, toDate: Option[String] = None, delta: Option[Int] = None, 
            pattern: String = "yyyy-MM-dd"): List[String]

    /**
     * Get HDFS compatible file paths to be used in tech stack like Spark. 
     * For ex: for S3 the file path is prefixed with s3n://<bucket>/<key> and for Azure blob storage it would be wasbs://<container-name>@<storage-account-name>.blob.core.windows.net/<key>/
     * 
     * @param container String - The container/bucket of the objects
     * @param objects List[Blob] - The Blob objects in the given container
     * 
     * @return List[String] - HDFS compatible file paths. 
     */
    def getPaths(container: String, objects: List[Blob]): List[String]

    /**
      * Get the blob object data.
      *
      * @param container String - The container/bucket of the file
      * @param objectKey String - The key/path of the blob object
      *
      * @return Array[String] - object data
      */
    def getObjectData(container: String, objectKey: String): Array[String]


    /**
      * Get the URI of the given prefix
      *
      * @param container
      * @param _prefix
      * @param isDirectory
      * @return String - URI of the given prefix
      */
    def getUri(container: String, _prefix: String, isDirectory: Option[Boolean] = Option(false)): String

}

File: cloud-storage-sdk-api/src/main/java/org/sunbird/cloud/storage/IStorageService.java
package org.sunbird.cloud.storage;

import org.sunbird.cloud.storage.model.Blob;
import org.sunbird.cloud.storage.model.DeleteTarget;

import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Cloud storage service interface providing operations for upload, download,
 * signed URL generation, deletion, copy, search, and metadata retrieval.
 *
 * <p>Implementations are CSP-specific (AWS, Azure, GCP, OCI) and discovered
 * via {@link StorageServiceFactory} using Java ServiceLoader.</p>
 */
public interface IStorageService extends AutoCloseable {

    // --- Upload operations ---

    /**
     * Upload a file or folder to cloud storage.
     *
     * @param container   the container/bucket name
     * @param file        the local file path
     * @param objectKey   the destination key/path in cloud storage
     * @param isDirectory whether to upload recursively as a directory
     * @param attempt     current attempt number (starting from 1)
     * @param retryCount  maximum number of retries before failing
     * @param ttl         time-to-live in seconds for the uploaded object (null for no expiry)
     * @return the URL of the uploaded file
     */
    String upload(String container, String file, String objectKey,
                  boolean isDirectory, int attempt, int retryCount, Integer ttl);

    default String upload(String container, String file, String objectKey) {
        return upload(container, file, objectKey, false, 1, 0, null);
    }

    /**
     * Upload a folder asynchronously.
     *
     * @param container  the container/bucket name
     * @param file       the local folder path
     * @param objectKey  the destination prefix in cloud storage
     * @param isPublic   whether uploaded files should have public read access
     * @param ttl        time-to-live in seconds (null for no expiry)
     * @param retryCount maximum number of retries (null for default)
     * @param attempt    current attempt number
     * @return future completing with a list of uploaded file URLs
     */
    CompletableFuture<List<String>> uploadFolder(String container, String file,
                                                  String objectKey, boolean isPublic,
                                                  Integer ttl, Integer retryCount, int attempt);

    default CompletableFuture<List<String>> uploadFolder(String container, String file,
                                                          String objectKey) {
        return uploadFolder(container, file, objectKey, false, null, null, 1);
    }

    /**
     * Put binary content directly as a blob in cloud storage.
     *
     * @param container   the container/bucket name
     * @param content     the byte array payload
     * @param objectKey   the destination key/path
     * @param isPublic    whether the object should have public read access
     * @param isDirectory whether the content represents a directory
     * @param ttl         time-to-live in seconds (null for no expiry)
     * @param retryCount  maximum number of retries (null for default)
     * @return the URL of the uploaded object
     */
    String put(String container, byte[] content, String objectKey,
               boolean isPublic, boolean isDirectory, Integer ttl, Integer retryCount);

    default String put(String container, byte[] content, String objectKey) {
        return put(container, content, objectKey, false, false, null, null);
    }

    // --- Signed URL operations ---

    /**
     * Get a pre-signed URL to access an object.
     *
     * @param container  the container/bucket name
     * @param objectKey  the object key
     * @param ttl        time-to-live in seconds for the URL (null for default)
     * @param permission "r" for read, "w" for write
     * @return the pre-signed URL
     */
    String getSignedURL(String container, String objectKey, Integer ttl, String permission);

    default String getSignedURL(String container, String objectKey) {
        return getSignedURL(container, objectKey, null, "r");
    }

    /**
     * Get a pre-signed URL with additional options (content type, extra params).
     *
     * @param container        the container/bucket name
     * @param objectKey        the object key
     * @param ttl              time-to-live in seconds (null for default)
     * @param permission       "r" for read, "w" for write
     * @param contentType      the content type of the object
     * @param additionalParams CSP-specific additional parameters (nullable)
     * @return the pre-signed URL
     */
    String getSignedURLV2(String container, String objectKey, Integer ttl,
                          String permission, String contentType,
                          Map<String, String> additionalParams);

    default String getSignedURLV2(String container, String objectKey) {
        return getSignedURLV2(container, objectKey, null, "r", "application/octet-stream", null);
    }

    // --- Download operations ---

    /**
     * Download a file or folder from cloud storage.
     *
     * @param container   the container/bucket name
     * @param objectKey   the object key to download
     * @param localPath   the local destination path
     * @param isDirectory whether to download recursively as a directory
     */
    void download(String container, String objectKey, String localPath, boolean isDirectory);

    default void download(String container, String objectKey, String localPath) {
        download(container, objectKey, localPath, false);
    }

    // --- Delete operations ---

    /**
     * Delete an object from cloud storage.
     *
     * @param container   the container/bucket name
     * @param objectKey   the object key to delete
     * @param isDirectory whether the object is a directory (recursive delete)
     */
    void deleteObject(String container, String objectKey, boolean isDirectory);

    default void deleteObject(String container, String objectKey) {
        deleteObject(container, objectKey, false);
    }

    /**
     * Delete multiple objects from cloud storage.
     *
     * @param container  the container/bucket name
     * @param objectKeys list of delete targets
     */
    void deleteObjects(String container, List<DeleteTarget> objectKeys);

    // --- Copy operations ---

    /**
     * Copy objects between containers or within the same container.
     *
     * @param fromContainer source container
     * @param fromKey       source object key/prefix
     * @param toContainer   destination container
     * @param toKey         destination object key/prefix
     * @param isDirectory   whether to copy recursively as a directory
     */
    void copyObjects(String fromContainer, String fromKey,
                     String toContainer, String toKey, boolean isDirectory);

    default void copyObjects(String fromContainer, String fromKey,
                             String toContainer, String toKey) {
        copyObjects(fromContainer, fromKey, toContainer, toKey, false);
    }

    // --- Archive operations ---

    /**
     * Extract an archive file on cloud storage to a destination folder.
     *
     * @param container the container/bucket name
     * @param objectKey the archive object key
     * @param toKey     the destination folder key
     */
    void extractArchive(String container, String objectKey, String toKey);

    // --- Object retrieval ---

    /**
     * Get blob object details.
     *
     * @param container   the container/bucket name
     * @param objectKey   the object key
     * @param withPayload whether to include the payload bytes
     * @return the blob object with metadata
     */
    Blob getObject(String container, String objectKey, boolean withPayload);

    default Blob getObject(String container, String objectKey) {
        return getObject(container, objectKey, false);
    }

    /**
     * Get object data as lines of text.
     *
     * @param container the container/bucket name
     * @param objectKey the object key
     * @return the object content as a list of lines
     */
    List<String> getObjectData(String container, String objectKey);

    /**
     * List blob objects for a given prefix.
     *
     * @param container   the container/bucket name
     * @param prefix      the object prefix to list
     * @param withPayload whether to include payloads
     * @return list of blob objects
     */
    List<Blob> listObjects(String container, String prefix, boolean withPayload);

    default List<Blob> listObjects(String container, String prefix) {
        return listObjects(container, prefix, false);
    }

    /**
     * List object keys for a given prefix.
     *
     * @param container the container/bucket name
     * @param prefix    the object prefix
     * @return list of object keys
     */
    List<String> listObjectKeys(String container, String prefix);

    // --- Search operations ---

    /**
     * Search for objects by date-prefixed paths.
     *
     * @param container the container/bucket name
     * @param prefix    the base prefix
     * @param fromDate  start date string (nullable)
     * @param toDate    end date string (nullable)
     * @param delta     number of days delta from toDate (nullable)
     * @param pattern   date format pattern (e.g., "yyyy-MM-dd")
     * @return list of matching blob objects
     */
    List<Blob> searchObjects(String container, String prefix,
                             String fromDate, String toDate,
                             Integer delta, String pattern);

    default List<Blob> searchObjects(String container, String prefix) {
        return searchObjects(container, prefix, null, null, null, "yyyy-MM-dd");
    }

    /**
     * Search for object keys by date-prefixed paths.
     *
     * @param container the container/bucket name
     * @param prefix    the base prefix
     * @param fromDate  start date string (nullable)
     * @param toDate    end date string (nullable)
     * @param delta     number of days delta from toDate (nullable)
     * @param pattern   date format pattern (e.g., "yyyy-MM-dd")
     * @return list of matching object keys
     */
    List<String> searchObjectKeys(String container, String prefix,
                                  String fromDate, String toDate,
                                  Integer delta, String pattern);

    default List<String> searchObjectKeys(String container, String prefix) {
        return searchObjectKeys(container, prefix, null, null, null, "yyyy-MM-dd");
    }

    // --- Path operations ---

    /**
     * Get HDFS-compatible file paths for use with Spark etc.
     *
     * @param container the container/bucket name
     * @param objects   list of blob objects
     * @return HDFS-compatible paths
     */
    List<String> getPaths(String container, List<Blob> objects);

    /**
     * Get the URI for a given prefix.
     *
     * @param container   the container/bucket name
     * @param prefix      the object prefix
     * @param isDirectory whether the prefix represents a directory
     * @return the URI string
     */
    String getUri(String container, String prefix, boolean isDirectory);

    default String getUri(String container, String prefix) {
        return getUri(container, prefix, false);
    }

    // --- Lifecycle ---

    @Override
    void close();
}
